util/notes/warnings.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notes
+
+import (
+	"errors"
+	"fmt"
+)
+
+type Warnings []error
+
+func (ws Warnings) Merge(notes Notes) Warnings {
+	nws := append(ws, notes.Warnings...)
+	for _, a := range notes.Annotations {
+		nws = append(nws, errors.New(a))
+	}
+	return nws
+}
+
+//nolint:revive // Ignore ST1012
+var (
+	RangeTooSmallWarning         = errors.New("Need at least 2 points to compute, perhaps time range is too small")
```

Comment:
I see three issues here:
1. The easy one: It should be "short", not "small", shouldn't it?
2. The moderately hard one: I think this should not be a warning, but an annotations (or info, or whatever nomenclature we end up with). I would say, a warning is when something definitely went wrong, but there are reasons to not fail the entire query because if it. However, it can be totally legitimate to have only 1 sample within the range (most prominently if the series has just started or stopped).
3. The really really hard nut to crack: This is just super prone to false positives (and even the idea with only reporting it for completely empty results doesn't help that much – but it makes things more complicated because now we can also have false negatives). As long as we only add this as an annotation, we could deem the false positives as acceptable. However, I think it will create more confusion than help.

So what would be the "right" conditions to issue this annotation, beyond having only one or even just zero samples in the range? My improvised attempt: All of the following must be true:

1. There must be no staleness marker in the range (because that means we are looking at the end of a series, which would explain that we don't see at least two samples).
2. There must be samples within a reasonable time _before_ the range so that there is even a chance to find more samples by extending the range. (This is hard in two ways: First, what is a reasonable time to look back? Currently, Prometheus doesn't know the intended scrape interval of a series. Maybe the lookback delta would be good as a start. Second, this requires a separate query to find out, which is a completely new flow.)
3. The last sample before the range must not be a staleness marker (because that hints towards an actual gap in the series, which is a case where it's expected to see only one sample in a range).

Reply:
It depends on how accurate we want this particular warning to be. If we just want to suggest a possible reason to the user why the query returns an empty result when doing a `rate` or something, and don't mind false positives/negatives (we could make it clear in the text that it might be wrong), then it's fine, but otherwise it seems like we'd need some complicated logic or to run additional queries just for a warning. In any case, I agree that we should just drop this for now, focus on adding the plumbing for warnings and the easy warnings first, and then tackle this in a future PR, so I've removed this warning and the associated logic from this one.

Reply:
OK, so let's postpone the range-to-short discussion for later. (Maybe somebody has found a proper solution for the "xrate" problem by then, and we don't have users anymore that accidentally pick a too short range…)